[SuiteSparse] Update and split into sub-packages

valgur opened the pull request:
SuiteSparse consists of a suite of mostly independent libraries, where each of them:
* has a distinct versioning scheme,
* varying license terms (e.g. some are GPL, while others are more permissive),
* installs a separate CMake config file (i.e. you would do `find_package(AMD)` to use `suitesparse-amd`),
* can have library-specific options, e.g. for CUDA support.

So I think splitting the package into its constituent sub-packages is well-justified.

Fixes #24252.

----

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

valgur committed:
[SuiteSparse] add suitesparse-config port

valgur committed:
[SuiteSparse] add suitesparse-amd

valgur committed:
[SuiteSparse] add suitesparse-btf

valgur committed:
[SuiteSparse] add suitesparse-camd

valgur committed:
[SuiteSparse] add suitesparse-ccolamd

valgur committed:
[SuiteSparse] add suitesparse-colamd

valgur committed:
[SuiteSparse] add suitesparse-ldl

valgur committed:
[SuiteSparse] add suitesparse-mongoose

valgur committed:
[SuiteSparse] add suitesparse-rbio

valgur committed:
[SuiteSparse] add suitesparse-spex

valgur committed:
[SuiteSparse] add suitesparse-graphblas

valgur committed:
[SuiteSparse] add suitesparse-lagraph

valgur committed:
[SuiteSparse] add suitesparse-cholmod

valgur committed:
[SuiteSparse] add suitesparse-klu

valgur committed:
[SuiteSparse] add suitesparse-umfpack

valgur committed:
[SuiteSparse] add suitesparse-paru

valgur committed:
[SuiteSparse] add suitesparse-spqr

valgur committed:
[SuiteSparse] add suitesparse-cxsparse

valgur committed:
[SuiteSparse] fix suitesparse-ldl license installation

valgur committed:
[SuiteSparse] update suitesparse meta-package

valgur committed:
[SuiteSparse] update Ceres and g2o to use specific SuiteSparse packages only

valgur committed:
[SuiteSparse] fix vcpkg.json formatting

valgur committed:
[SuiteSparse] update version database

Neumann-A commented:
> (i.e. you would do find_package(AMD) to use suitesparse-amd),

That looks like a no go. Is there an equivalent of `find_package(Suitesparse COPMPONENTS AMD)` ?

valgur commented:
> > (i.e. you would do find_package(AMD) to use suitesparse-amd),
> 
> That looks like a no go. Is there an equivalent of `find_package(Suitesparse COPMPONENTS AMD)` ?

AFAIK SuiteSparse exports a `SuiteSparse` CMake module or config.
Both Ceres and g2o (the only users of the suitesparse port, currently) define a custom [`FindSuiteSparse.cmake`](https://github.com/ceres-solver/ceres-solver/blob/master/cmake/FindSuiteSparse.cmake).

valgur committed:
[SuiteSparse] fix suitesparse-config fixup

valgur committed:
[SuiteSparse] vdb

Neumann-A commented:
Yeah I checked upstream. Their `find_package` logic is flawed but nothing for you to fix. There is a Suitesparse_config package but that won't help much. 

They should really install Suitesparse-config.cmake and use the component approach instead of just taking common abbreviations. SuitesparseAMD-config.cmake would be much better compared to just AMD-config.cmake.

valgur committed:
[SuiteSparse] set DISABLE_PARALLEL_CONFIGURE

Parallel configure fails due to race conditions with configure_file().

valgur committed:
[SuiteSparse] vdb

valgur commented:
@Neumann-A `find_package(LAPACK REQUIRED)` fails despite `lapack` being listed as a dependency:
```
CMake Error at /mnt/vcpkg-ci/installed/x64-linux/share/lapack/FindLAPACK.cmake:522 (message):
  A required library with LAPACK API not found.  Please specify library location.
```
Do you have any ideas what might be causing this? I can reproduce it locally as well.

Neumann-A commented:
> find_package(LAPACK REQUIRED) fails despite lapack being listed as a dependency:

Can have numerous reasons. probably somebody messing with BLA_VENDOR again.

BillyONeal commented:
There are already several failures and the Windows lab is very behind so I cancelled this one

valgur committed:
[SuiteSparse] disable OpenMP by default

valgur committed:
[SuiteSparse] update vdb

MonicaLiu0311 commented on ports/suitesparse/vcpkg.json:
```diff
@@ -1,19 +1,113 @@
-  "description": "A suite of sparse matrix algorithms. Also provides libcxsparse.",
-  "homepage": "http://suitesparse.com",
+  "version-semver": "7.7.0",
+  "port-version": 3,
```
```suggestion
```

MonicaLiu0311 replied:
There are two situations when modifying files under `vcpkg/ports/portName`:
1. If a version is released upstream, update the `"version"` field in `vcpkg.json` and count `"port-version"` from 0 (port-version is a field that records changes specific to vcpkg, 0 generally means deleting "port-version" line). Reference: [38214/files](https://github.com/microsoft/vcpkg/pull/38214/files).
2. If the upstream has not released a new version, then after modifying other files, please accumulate the `"port-version"` value to record this modification. Reference: [37966/files](https://github.com/microsoft/vcpkg/pull/37966/files).

Please read the documentation: [maintainer guide#versioning](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md#versioning).

valgur committed:
[SuiteSparse] fix port-version

valgur committed:
[SuiteSparse] update openmp in suitesparse/vcpkg.json

valgur committed:
[SuiteSparse] workaround for find_package(LAPACK) issues due to BLA_VENDOR

valgur committed:
[Ceres] add compatibility with SuiteSparse v7

Commit:
update git-tree

MonicaLiu0311 commented:
At present, the version information of your commit is inconsistent with your local one:
```
##[error]Found the following errors:
##[error]versions/s-/suitesparse.json(2,): error : while parsing versions for suitesparse from version/s-/suitesparse.json


##[error]versions/s-/suitesparse-cholmod.json(2,): error : while parsing versions for suitesparse-cholmod from version/s-/suitesparse-cholmod.json
5.2.1 is declared with c27e51b093a40997bef22054c781959c5f5e46c8, but the local port has a different SHA cddaa587403e37f0a1193f0ee3e96b3f6da9bfe2.

##[error]versions/s-/suitesparse-config.json(2,): error : while parsing versions for suitesparse-config from version/s-/suitesparse-config.json
7.7.0 is declared with e614c9e5b1e01d8121e509470339fe1f8d2f33cc, but the local port has a different SHA 26262a5f8296b1227a70c05d2d94df52e01ff3ad.

##[error]versions/s-/suitesparse-spqr.json(2,): error : while parsing versions for suitesparse-spqr from version/s-/suitesparse-spqr.json
4.3.3 is declared with 85e1a30febea3dd072a24116b93b62ed84f683e1, but the local port has a different SHA ff8e4cfa714edffd5fda753858f4624a6ff1ec42.
```
Please run `./vcpkg x-add-version testPort [--overwrite-version]` before each push:
1. For any file modification under the `ports/testPort/` folder, please commit after modification
2. Run `./vcpkg x-add-version portName` to generate a series of version according to your commit info (Add `--overwrite-version` when generating other than the first time)
3. Commit verison information again
4. Push together

`--overwrite-version`
After the first modification, please run `./vcpkg x-add-version testPort` to generate the corresponding json; for subsequent modifications of the same version, please use `./vcpkg x-add-version portName --overwrite-version` to update git-tree.

[maintainer guide#versioning](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md#versioning).


What I did locally:
![image](https://github.com/microsoft/vcpkg/assets/110024546/bb089e42-9d68-414f-9784-7def9410b7d5)

valgur committed:
[Ceres] update vdb

valgur committed:
[Ceres] add missing suitesparse-cxsparse dependency

valgur committed:
[Ceres] update vdb

valgur committed:
[Ceres] fix CeresConfig.cmake.in

valgur committed:
[Ceres] update vdb

valgur committed:
[SuiteSparse] add missing openmp feature to lagraph

valgur committed:
[SuiteSparse] update vdb

valgur committed:
[theia] add support for SuiteSparse v7

valgur committed:
[theia] update vdb

valgur committed:
[Theia] update and merge patches, drop default [viewer] from openimageio

valgur committed:
[Theia] update vdb

valgur committed:
[SuiteSparse] handle suitesparse_mongoose tool

valgur committed:
[SuiteSparse] update vdb

valgur committed:
[SuiteSparse] fix GraphBLAS JIT cache dir handling

valgur committed:
[SuiteSparse] update vdb

valgur committed:
[SuiteSparse] disable complex data types for ARM android

valgur committed:
vdb

MonicaLiu0311 commented:
@valgur Is there any new progress?

valgur commented:
> @valgur Is there any new progress?

I have been focusing on #39389 lately. Since that one is now complete and ready as far as I'm concerned, I'll be looking at this one next. The LAPACK and BLAS issues look quite tricky to solve, though, so it might take some time.

valgur committed:
[suitesparse] fix BLAS linking on Windows

valgur committed:
[suitesparse-cxsparse] complex is not available on MSVC

valgur committed:
[suitesparse] fix BLAS linking in components using it transitively

valgur committed:
vdb

valgur committed:
[suitesparse-graphblas] disable on android

GrpahBLAS has a strict requirement for complex math support, which Android NDK seems to lack.

valgur committed:
vdb

valgur commented on ports/suitesparse-config/SuiteSparseBLAS.cmake:
Not super happy about this workaround for BLAS and LAPACK handling, but it works well enough with current Vcpkg BLAS and LAPACK setup. This is a replacement for https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/dev/SuiteSparse_config/cmake_modules/SuiteSparseBLAS.cmake, which modifies `BLA_VENDOR` and breaks LAPACK detection as a side-effect. Maybe the more generic BLAS handling in the project can be restored in a follow-up PR, if there's interest.

valgur commented on ports/suitesparse-config/SuiteSparseBLAS.cmake:
```diff
@@ -0,0 +1,9 @@
+set(BLA_SIZEOF_INTEGER 4)
+set(SuiteSparse_BLAS_integer int32_t)
+
+if(WIN32)
+    # OpenBLAS includes an underscore suffix on Windows for all of its symbols.
+    # This is not detected automatically by SuiteSparse or FindBLAS and needs to be set manually.
+    add_compile_definitions(BLAS64__SUFFIX=_)
+endif()
```
A `check_symbol_exists()` call ro check for `dgemm`/`dgemm_` would be a more robust and generic approach, but this at least works for now.

MonicaLiu0311 commented:
* **Feature testing**

    ceres
    -
    Features `cxsparse,suitesparse` are tested successfully in the following triplet:
    ```
    x86-windows
    x64-windows
    x64-windows-static
    ```

    suitesparse-config & suitesparse-klu & suitesparse-lagraph & suitesparse-mongoose & suitesparse-spex & suitesparse-umfpack
    -
    All features are tested successfully in the following triplet:
    ```
    x86-windows
    x64-windows
    x64-windows-static
    ```

    suitesparse-graphblas
    -
    All features are tested successfully in the following triplet:
    ```
    x64-windows
    x64-windows-static
    ```

    suitesparse-cholmod
    -
    Features `matrixops,modify,openmp,partition,supernodal` are tested successfully in the following triplet:
    ```
    x86-windows
    x64-windows
    x64-windows-static
    ```

    suitesparse
    -
    Feature `gpl` is tested successfully in the following triplet:
    ```
    x86-windows
    x64-windows
    x64-windows-static
    ```

MonicaLiu0311 commented:
* **Usage testing**

    The usage test passed on `x64-windows` (header files found).

    <details>
    <summary><b>usage</b></summary>
    <pre>

    suitesparse-amd provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(AMD CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::AMD)

    suitesparse-camd provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(CAMD CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::CAMD)

    suitesparse-ccolamd provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(CCOLAMD CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::CCOLAMD)

    suitesparse-colamd provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(COLAMD CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::COLAMD)

    suitesparse-config provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(SuiteSparse_config CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::SuiteSparseConfig)

    suitesparse-btf provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(BTF CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::BTF)

    suitesparse-graphblas provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(GraphBLAS CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::GraphBLAS)

    suitesparse-lagraph provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(LAGraph CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::LAGraph SuiteSparse::LAGraphX)

    suitesparse-ldl provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(LDL CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::LDL)

    suitesparse-mongoose provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(SuiteSparse_Mongoose CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::Mongoose)

    suitesparse-rbio provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(RBio CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::RBio)

    suitesparse-spex provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(SPEX CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::SPEX)

    suitesparse-cholmod provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(CHOLMOD CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::CHOLMOD)

    suitesparse-cxsparse provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(CXSparse CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::CXSparse)

    suitesparse-spqr provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(SPQR CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::SPQR)

    ceres provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(Ceres CONFIG REQUIRED)
      target_link_libraries(main PRIVATE Ceres::ceres)

    suitesparse-klu provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(KLU CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::KLU)

    suitesparse-umfpack provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(UMFPACK CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::UMFPACK)

    suitesparse-paru provides CMake targets:

      # this is heuristically generated, and may not be correct
      find_package(ParU CONFIG REQUIRED)
      target_link_libraries(main PRIVATE SuiteSparse::ParU)
    </pre></details>

MonicaLiu0311 commented:
suitesparse-paru & suitesparse
-
When testing `openmp` feature, the following error occurs:
```
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: G:/suitesparse/downloads/tools/cmake-3.29.2-windows/cmake-3.29.2-windows-i386/bin/cmake.exe G:/suitesparse/buildtrees/suitesparse-paru/src/v7.7.0-b453f37041.clean/ParU -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=G:/suitesparse/packages/suitesparse-paru_x64-windows/debug -DFETCHCONTENT_FULLY_DISCONNECTED=ON -DBUILD_STATIC_LIBS=0 -DSUITESPARSE_USE_CUDA=OFF -DSUITESPARSE_USE_STRICT=ON -DSUITESPARSE_USE_FORTRAN=OFF -DSUITESPARSE_DEMOS=OFF -DSUITESPARSE_USE_OPENMP=ON "-DCMAKE_MAKE_PROGRAM=C:/Program Files/Microsoft Visual Studio/2022/Enterprise/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -DBUILD_SHARED_LIBS=ON -DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=G:/suitesparse/scripts/toolchains/windows.cmake -DVCPKG_TARGET_TRIPLET=x64-windows -DVCPKG_SET_CHARSET_FLAG=ON -DVCPKG_PLATFORM_TOOLSET=v143 -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE -DCMAKE_VERBOSE_MAKEFILE=ON -DVCPKG_APPLOCAL_DEPS=OFF -DCMAKE_TOOLCHAIN_FILE=G:/suitesparse/scripts/buildsystems/vcpkg.cmake -DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON -DVCPKG_CXX_FLAGS= -DVCPKG_CXX_FLAGS_RELEASE= -DVCPKG_CXX_FLAGS_DEBUG= -DVCPKG_C_FLAGS= -DVCPKG_C_FLAGS_RELEASE= -DVCPKG_C_FLAGS_DEBUG= -DVCPKG_CRT_LINKAGE=dynamic -DVCPKG_LINKER_FLAGS= -DVCPKG_LINKER_FLAGS_RELEASE= -DVCPKG_LINKER_FLAGS_DEBUG= -DVCPKG_TARGET_ARCHITECTURE=x64 -DCMAKE_INSTALL_LIBDIR:STRING=lib -DCMAKE_INSTALL_BINDIR:STRING=bin -D_VCPKG_ROOT_DIR=G:/suitesparse -D_VCPKG_INSTALLED_DIR=G:/suitesparse/installed -DVCPKG_MANIFEST_INSTALL=OFF
    Working Directory: G:/suitesparse/buildtrees/suitesparse-paru/x64-windows-dbg
    Error code: 1
    See logs for more information:
      G:\suitesparse\buildtrees\suitesparse-paru\config-x64-windows-dbg-CMakeCache.txt.log
      G:\suitesparse\buildtrees\suitesparse-paru\config-x64-windows-dbg-out.log
      G:\suitesparse\buildtrees\suitesparse-paru\config-x64-windows-dbg-err.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:276 (vcpkg_execute_required_process)
  ports/suitesparse-paru/portfile.cmake:26 (vcpkg_cmake_configure)
  scripts/ports.cmake:175 (include)
```
**config-x64-windows-dbg-err.log:**
```
CMake Error at CMakeLists.txt:92 (message):
  OpenMP 4.5 required for ParU but not found
```

suitesparse-cxsparse
-
When testing `complex` feature, the following error occurs:
```
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: G:/suitesparse/downloads/tools/cmake-3.29.2-windows/cmake-3.29.2-windows-i386/bin/cmake.exe --build . --config Debug --target install -- -v -j9
    Working Directory: G:/suitesparse/buildtrees/suitesparse-cxsparse/x64-windows-dbg
    See logs for more information:
      G:\suitesparse\buildtrees\suitesparse-cxsparse\install-x64-windows-dbg-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/suitesparse-cxsparse/portfile.cmake:29 (vcpkg_cmake_install)
  scripts/ports.cmake:175 (include)
```
**install-x64-windows-dbg-out.log:**
```
[1/208] C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1440~1.338\bin\Hostx64\x64\cl.exe   -DCXSparse_EXPORTS -D_CRT_SECURE_NO_WARNINGS -IG:\suitesparse\buildtrees\suitesparse-cxsparse\x64-windows-dbg -IG:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse -IG:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Source -IG:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Include -IG:\suitesparse\installed\x64-windows\include\suitesparse /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  -std:c11 -MDd /showIncludes /FoCMakeFiles\CXSparse.dir\Source\cs_add.c.obj /FdCMakeFiles\CXSparse.dir\ /FS -c G:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Source\cs_add.c
FAILED: CMakeFiles/CXSparse.dir/Source/cs_add.c.obj 
C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1440~1.338\bin\Hostx64\x64\cl.exe   -DCXSparse_EXPORTS -D_CRT_SECURE_NO_WARNINGS -IG:\suitesparse\buildtrees\suitesparse-cxsparse\x64-windows-dbg -IG:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse -IG:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Source -IG:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Include -IG:\suitesparse\installed\x64-windows\include\suitesparse /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /MP  /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  -std:c11 -MDd /showIncludes /FoCMakeFiles\CXSparse.dir\Source\cs_add.c.obj /FdCMakeFiles\CXSparse.dir\ /FS -c G:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Source\cs_add.c
G:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Include\cs.h(357): error C2143: syntax error: missing ';' before '*'
G:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Include\cs.h(357): error C2059: syntax error: '*'
G:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Include\cs.h(359): error C2059: syntax error: '}'
G:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Include\cs.h(361): error C2143: syntax error: missing '{' before '*'
G:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Include\cs.h(361): error C2143: syntax error: missing ')' before '*'
G:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Include\cs.h(361): error C2143: syntax error: missing '{' before '*'
G:\suitesparse\buildtrees\suitesparse-cxsparse\src\v7.7.0-0fc10962f1.clean\CXSparse\Include\cs.h(361): warning C4228: nonstandard extension used: qualifiers after comma in declarator list are ignored
```

suitesparse-cholmod & suitesparse-spqr
-
When testing `cuda,cuda-redist` feature, the following error occurs:
```
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: G:/suitesparse/downloads/tools/cmake-3.29.2-windows/cmake-3.29.2-windows-i386/bin/cmake.exe G:/suitesparse/buildtrees/suitesparse-spqr/src/v7.7.0-0fc10962f1.clean/SPQR -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=G:/suitesparse/packages/suitesparse-spqr_x64-windows/debug -DFETCHCONTENT_FULLY_DISCONNECTED=ON -DBUILD_STATIC_LIBS=0 -DSUITESPARSE_USE_CUDA=ON -DSPQR_USE_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=all-major -DSUITESPARSE_USE_STRICT=ON -DSUITESPARSE_USE_FORTRAN=OFF -DSUITESPARSE_DEMOS=OFF "-DCMAKE_MAKE_PROGRAM=C:/Program Files/Microsoft Visual Studio/2022/Enterprise/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -DBUILD_SHARED_LIBS=ON -DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=G:/suitesparse/scripts/toolchains/windows.cmake -DVCPKG_TARGET_TRIPLET=x64-windows -DVCPKG_SET_CHARSET_FLAG=ON -DVCPKG_PLATFORM_TOOLSET=v143 -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE -DCMAKE_VERBOSE_MAKEFILE=ON -DVCPKG_APPLOCAL_DEPS=OFF -DCMAKE_TOOLCHAIN_FILE=G:/suitesparse/scripts/buildsystems/vcpkg.cmake -DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON -DVCPKG_CXX_FLAGS= -DVCPKG_CXX_FLAGS_RELEASE= -DVCPKG_CXX_FLAGS_DEBUG= -DVCPKG_C_FLAGS= -DVCPKG_C_FLAGS_RELEASE= -DVCPKG_C_FLAGS_DEBUG= -DVCPKG_CRT_LINKAGE=dynamic -DVCPKG_LINKER_FLAGS= -DVCPKG_LINKER_FLAGS_RELEASE= -DVCPKG_LINKER_FLAGS_DEBUG= -DVCPKG_TARGET_ARCHITECTURE=x64 -DCMAKE_INSTALL_LIBDIR:STRING=lib -DCMAKE_INSTALL_BINDIR:STRING=bin -D_VCPKG_ROOT_DIR=G:/suitesparse -D_VCPKG_INSTALLED_DIR=G:/suitesparse/installed -DVCPKG_MANIFEST_INSTALL=OFF
    Working Directory: G:/suitesparse/buildtrees/suitesparse-spqr/x64-windows-dbg
    Error code: 1
    See logs for more information:
      G:\suitesparse\buildtrees\suitesparse-spqr\config-x64-windows-dbg-CMakeCache.txt.log
      G:\suitesparse\buildtrees\suitesparse-spqr\config-x64-windows-dbg-out.log
      G:\suitesparse\buildtrees\suitesparse-spqr\config-x64-windows-dbg-err.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:276 (vcpkg_execute_required_process)
  ports/suitesparse-spqr/portfile.cmake:29 (vcpkg_cmake_configure)
  scripts/ports.cmake:175 (include)
```
**config-x64-windows-dbg-err.log:**
```
CMake Error at G:/suitesparse/buildtrees/suitesparse-spqr/src/v7.7.0-0fc10962f1.clean/SuiteSparse_config/cmake_modules/SuiteSparsePolicy.cmake:363 (message):
  CUDA required for SuiteSparse but not found
Call Stack (most recent call first):
  CMakeLists.txt:41 (include)
```


suitesparse-graphblas:x86-windows
-
When installing `suitesparse-graphblas:x86-windows`, the following error occurs:
```
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: G:/suitesparse/downloads/tools/cmake-3.29.2-windows/cmake-3.29.2-windows-i386/bin/cmake.exe --build . --config Debug --target install -- -v -j9
    Working Directory: G:/suitesparse/buildtrees/suitesparse-graphblas/x86-windows-dbg
    See logs for more information:
      G:\suitesparse\buildtrees\suitesparse-graphblas\install-x86-windows-dbg-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/suitesparse-graphblas/portfile.cmake:35 (vcpkg_cmake_install)
  scripts/ports.cmake:175 (include)
```
**install-x86-windows-dbg-out.log:**
```
[797/798] C:\Windows\system32\cmd.exe /C "C:\Windows\system32\cmd.exe /C "G:\suitesparse\downloads\tools\cmake-3.29.2-windows\cmake-3.29.2-windows-i386\bin\cmake.exe -E __create_def G:\suitesparse\buildtrees\suitesparse-graphblas\x86-windows-dbg\CMakeFiles\GraphBLAS.dir\.\exports.def G:\suitesparse\buildtrees\suitesparse-graphblas\x86-windows-dbg\CMakeFiles\GraphBLAS.dir\.\exports.def.objs && cd G:\suitesparse\buildtrees\suitesparse-graphblas\x86-windows-dbg" && G:\suitesparse\downloads\tools\cmake-3.29.2-windows\cmake-3.29.2-windows-i386\bin\cmake.exe -E vs_link_dll --intdir=CMakeFiles\GraphBLAS.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1440~1.338\bin\Hostx64\x86\link.exe  @CMakeFiles\GraphBLAS.rsp  /out:graphblas.dll /implib:graphblas.lib /pdb:graphblas.pdb /dll /version:9.2 /machine:X86 /nologo    /debug /INCREMENTAL  /DEF:CMakeFiles\GraphBLAS.dir\.\exports.def && cd ."
FAILED: graphblas.dll graphblas.lib 
C:\Windows\system32\cmd.exe /C "C:\Windows\system32\cmd.exe /C "G:\suitesparse\downloads\tools\cmake-3.29.2-windows\cmake-3.29.2-windows-i386\bin\cmake.exe -E __create_def G:\suitesparse\buildtrees\suitesparse-graphblas\x86-windows-dbg\CMakeFiles\GraphBLAS.dir\.\exports.def G:\suitesparse\buildtrees\suitesparse-graphblas\x86-windows-dbg\CMakeFiles\GraphBLAS.dir\.\exports.def.objs && cd G:\suitesparse\buildtrees\suitesparse-graphblas\x86-windows-dbg" && G:\suitesparse\downloads\tools\cmake-3.29.2-windows\cmake-3.29.2-windows-i386\bin\cmake.exe -E vs_link_dll --intdir=CMakeFiles\GraphBLAS.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1440~1.338\bin\Hostx64\x86\link.exe  @CMakeFiles\GraphBLAS.rsp  /out:graphblas.dll /implib:graphblas.lib /pdb:graphblas.pdb /dll /version:9.2 /machine:X86 /nologo    /debug /INCREMENTAL  /DEF:CMakeFiles\GraphBLAS.dir\.\exports.def && cd ."
LINK Pass 1: command "C:\PROGRA~1\MICROS~2\2022\ENTERP~1\VC\Tools\MSVC\1440~1.338\bin\Hostx64\x86\link.exe @CMakeFiles\GraphBLAS.rsp /out:graphblas.dll /implib:graphblas.lib /pdb:graphblas.pdb /dll /version:9.2 /machine:X86 /nologo /debug /INCREMENTAL /DEF:CMakeFiles\GraphBLAS.dir\.\exports.def /MANIFEST /MANIFESTFILE:CMakeFiles\GraphBLAS.dir/intermediate.manifest CMakeFiles\GraphBLAS.dir/manifest.res" failed (exit code 1120) with the following output:
   Creating library graphblas.lib and object graphblas.exp
GB_AxB_saxpy3_generic_secondj32.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_generic_secondj64.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_generic_unflipped.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_generic_firstj32.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_generic_firstj64.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_generic_flipped.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_generic_second.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB__any_pair_iso.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_generic_first.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_generic_firsti32.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_generic_firsti64.c.obj : error LNK2001: unresolved external symbol __InterlockedOr64
GB_AxB_saxpy3_sym_nsb.c.obj : error LNK2019: unresolved external symbol __InterlockedExchange64 referenced in function _GB_AxB_saxpy3_sym_nsb$omp$1
```

valgur committed:
[suitesparse-paru] MSVC OpenMP is not compatible (v2.0 < required v4.5)

valgur committed:
[suitesparse-cxsparse] strictly disable 'complex' on Windows

valgur committed:
[suitesparse] disable CUDA feature on Windows

Don't have a good enough setup to debug this configuration currently.
Works locally on Linux, though.

valgur committed:
[suitesparse-graphblas] disable x86 Windows

unresolved external symbol __InterlockedOr64

valgur committed:
[suitesparse] fix vcpkg.json formatting

valgur committed:
vdb

valgur committed:
[suitesparse-cxsparse] don't enable 'complex' on Windows

valgur committed:
[suitesparse] fix 'cuda' feature support

'matrixops' feature is required in CHOLMOD and SPQR was not packaging two CMake config files correctly

valgur committed:
vdb

valgur committed:
[colmap] add a missing METIS dependency

Used to be provided transitively by SuiteSparse before.

valgur committed:
vdb

valgur commented:
Fixed the incomplete CUDA support. Builds without issues now for `x64-windows` with

```
vcpkg install suitesparse-cholmod[cuda-redist] suitesparse-spqr[cuda-redist] ceres[suitesparse] g2o colmap[cuda-redist] theia
```

MonicaLiu0311 commented:
> suitesparse-cholmod & suitesparse-spqr
> -
> When testing `cuda,cuda-redist` feature, the following error occurs:
> ```
> CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
>     Command failed: G:/suitesparse/downloads/tools/cmake-3.29.2-windows/cmake-3.29.2-windows-i386/bin/cmake.exe G:/suitesparse/buildtrees/suitesparse-spqr/src/v7.7.0-0fc10962f1.clean/SPQR -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=G:/suitesparse/packages/suitesparse-spqr_x64-windows/debug -DFETCHCONTENT_FULLY_DISCONNECTED=ON -DBUILD_STATIC_LIBS=0 -DSUITESPARSE_USE_CUDA=ON -DSPQR_USE_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=all-major -DSUITESPARSE_USE_STRICT=ON -DSUITESPARSE_USE_FORTRAN=OFF -DSUITESPARSE_DEMOS=OFF "-DCMAKE_MAKE_PROGRAM=C:/Program Files/Microsoft Visual Studio/2022/Enterprise/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -DBUILD_SHARED_LIBS=ON -DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=G:/suitesparse/scripts/toolchains/windows.cmake -DVCPKG_TARGET_TRIPLET=x64-windows -DVCPKG_SET_CHARSET_FLAG=ON -DVCPKG_PLATFORM_TOOLSET=v143 -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE -DCMAKE_VERBOSE_MAKEFILE=ON -DVCPKG_APPLOCAL_DEPS=OFF -DCMAKE_TOOLCHAIN_FILE=G:/suitesparse/scripts/buildsystems/vcpkg.cmake -DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON -DVCPKG_CXX_FLAGS= -DVCPKG_CXX_FLAGS_RELEASE= -DVCPKG_CXX_FLAGS_DEBUG= -DVCPKG_C_FLAGS= -DVCPKG_C_FLAGS_RELEASE= -DVCPKG_C_FLAGS_DEBUG= -DVCPKG_CRT_LINKAGE=dynamic -DVCPKG_LINKER_FLAGS= -DVCPKG_LINKER_FLAGS_RELEASE= -DVCPKG_LINKER_FLAGS_DEBUG= -DVCPKG_TARGET_ARCHITECTURE=x64 -DCMAKE_INSTALL_LIBDIR:STRING=lib -DCMAKE_INSTALL_BINDIR:STRING=bin -D_VCPKG_ROOT_DIR=G:/suitesparse -D_VCPKG_INSTALLED_DIR=G:/suitesparse/installed -DVCPKG_MANIFEST_INSTALL=OFF
>     Working Directory: G:/suitesparse/buildtrees/suitesparse-spqr/x64-windows-dbg
>     Error code: 1
>     See logs for more information:
>       G:\suitesparse\buildtrees\suitesparse-spqr\config-x64-windows-dbg-CMakeCache.txt.log
>       G:\suitesparse\buildtrees\suitesparse-spqr\config-x64-windows-dbg-out.log
>       G:\suitesparse\buildtrees\suitesparse-spqr\config-x64-windows-dbg-err.log
> 
> Call Stack (most recent call first):
>   installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:276 (vcpkg_execute_required_process)
>   ports/suitesparse-spqr/portfile.cmake:29 (vcpkg_cmake_configure)
>   scripts/ports.cmake:175 (include)
> ```
> **config-x64-windows-dbg-err.log:**
> ```
> CMake Error at G:/suitesparse/buildtrees/suitesparse-spqr/src/v7.7.0-0fc10962f1.clean/SuiteSparse_config/cmake_modules/SuiteSparsePolicy.cmake:363 (message):
>   CUDA required for SuiteSparse but not found
> Call Stack (most recent call first):
>   CMakeLists.txt:41 (include)
> ```

I failed to install on both Windows and Linux, which version of CUDA did you install locally?
![image](https://github.com/user-attachments/assets/703ee788-e161-47af-b87f-146c944e6c88)

MonicaLiu0311 commented:
> suitesparse-cholmod & suitesparse-spqr
> -
> When testing `cuda,cuda-redist` feature, the following error occurs:
> ```
> CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
>     Command failed: G:/suitesparse/downloads/tools/cmake-3.29.2-windows/cmake-3.29.2-windows-i386/bin/cmake.exe G:/suitesparse/buildtrees/suitesparse-spqr/src/v7.7.0-0fc10962f1.clean/SPQR -G Ninja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=G:/suitesparse/packages/suitesparse-spqr_x64-windows/debug -DFETCHCONTENT_FULLY_DISCONNECTED=ON -DBUILD_STATIC_LIBS=0 -DSUITESPARSE_USE_CUDA=ON -DSPQR_USE_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=all-major -DSUITESPARSE_USE_STRICT=ON -DSUITESPARSE_USE_FORTRAN=OFF -DSUITESPARSE_DEMOS=OFF "-DCMAKE_MAKE_PROGRAM=C:/Program Files/Microsoft Visual Studio/2022/Enterprise/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -DBUILD_SHARED_LIBS=ON -DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=G:/suitesparse/scripts/toolchains/windows.cmake -DVCPKG_TARGET_TRIPLET=x64-windows -DVCPKG_SET_CHARSET_FLAG=ON -DVCPKG_PLATFORM_TOOLSET=v143 -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON -DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE -DCMAKE_VERBOSE_MAKEFILE=ON -DVCPKG_APPLOCAL_DEPS=OFF -DCMAKE_TOOLCHAIN_FILE=G:/suitesparse/scripts/buildsystems/vcpkg.cmake -DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON -DVCPKG_CXX_FLAGS= -DVCPKG_CXX_FLAGS_RELEASE= -DVCPKG_CXX_FLAGS_DEBUG= -DVCPKG_C_FLAGS= -DVCPKG_C_FLAGS_RELEASE= -DVCPKG_C_FLAGS_DEBUG= -DVCPKG_CRT_LINKAGE=dynamic -DVCPKG_LINKER_FLAGS= -DVCPKG_LINKER_FLAGS_RELEASE= -DVCPKG_LINKER_FLAGS_DEBUG= -DVCPKG_TARGET_ARCHITECTURE=x64 -DCMAKE_INSTALL_LIBDIR:STRING=lib -DCMAKE_INSTALL_BINDIR:STRING=bin -D_VCPKG_ROOT_DIR=G:/suitesparse -D_VCPKG_INSTALLED_DIR=G:/suitesparse/installed -DVCPKG_MANIFEST_INSTALL=OFF
>     Working Directory: G:/suitesparse/buildtrees/suitesparse-spqr/x64-windows-dbg
>     Error code: 1
>     See logs for more information:
>       G:\suitesparse\buildtrees\suitesparse-spqr\config-x64-windows-dbg-CMakeCache.txt.log
>       G:\suitesparse\buildtrees\suitesparse-spqr\config-x64-windows-dbg-out.log
>       G:\suitesparse\buildtrees\suitesparse-spqr\config-x64-windows-dbg-err.log
> 
> Call Stack (most recent call first):
>   installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:276 (vcpkg_execute_required_process)
>   ports/suitesparse-spqr/portfile.cmake:29 (vcpkg_cmake_configure)
>   scripts/ports.cmake:175 (include)
> ```
> **config-x64-windows-dbg-err.log:**
> ```
> CMake Error at G:/suitesparse/buildtrees/suitesparse-spqr/src/v7.7.0-0fc10962f1.clean/SuiteSparse_config/cmake_modules/SuiteSparsePolicy.cmake:363 (message):
>   CUDA required for SuiteSparse but not found
> Call Stack (most recent call first):
>   CMakeLists.txt:41 (include)
> ```

I failed to install on both Windows and Linux, which version of CUDA did you install locally?
![image](https://github.com/user-attachments/assets/703ee788-e161-47af-b87f-146c944e6c88)

valgur commented:
@MonicaLiu0311 I used the latest CUDA 12.5 on both Windows and Linux.

MonicaLiu0311 commented on ports/suitesparse-graphblas/vcpkg.json:
```diff
@@ -0,0 +1,29 @@
+  "description": "SuiteSparse:GraphBLAS: graph algorithms in the language of linear algebra",
+  "homepage": "https://people.engr.tamu.edu/davis/GraphBLAS.html",
+  "license": "Apache-2.0",
+  "supports": "!android & !(windows & x86)",
```
`"support"` must be consistent with upstream. I did not find any [upstream documentation](https://github.com/DrTimothyAldenDavis/GraphBLAS/blob/v9.2.0/GraphBLAS/README.md) on `x86-windows` not being supported. If you cannot resolve errors on this triplet, please add it to [ci.baseline.txt](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt#L1118).

Installation successful:
```
suitesparse-graphblas"[compact]":arm64-osx
suitesparse-graphblas"[*]":x64-android
```

Installation failed:
```
suitesparse-graphblas"[openmp]":arm64-osx
suitesparse-graphblas:x86-windows
```

valgur replied:
Looks like I'll have to report the issue upstream. Linking fails for x86 Debug builds only due to
- `__InterlockedExchange64`
- `__InterlockedIncrement64`
- `__InterlockedOr64`
not being found. Static/dynamic CRT linkage does not appear to have any effect.

dg0yt replied:
> Linking fails for x86 Debug builds only due to
> 
>     * `__InterlockedExchange64`
> 
>     * `__InterlockedIncrement64`
> 
>     * `__InterlockedOr64`
> 
> 
> not being found.

Not supported for x86 Windows:
https://learn.microsoft.com/en-us/cpp/intrinsics/interlockedexchange-intrinsic-functions?view=msvc-170#requirements
https://learn.microsoft.com/en-us/cpp/intrinsics/interlockedincrement-intrinsic-functions?view=msvc-170#requirements
https://learn.microsoft.com/en-us/cpp/intrinsics/interlockedor-intrinsic-functions?view=msvc-170#requirements

valgur committed:
[suitesparse-graphblas] add x86-windows to ci.baseline.txt

valgur commented:
I added a `ci.baseline.txt` exception for `suitesparse-graphblas`, but the CI pipeline now fails with:

> Check For Common Mistakes / Check (pull_request): GitHub Actions has encountered an internal error when running your job.

@MonicaLiu0311 Do you know what might have gone wrong there? The change to `ci.baseline.txt` looks ok to me.

valgur committed:
Bump port-versions

valgur committed:
update vdb

MonicaLiu0311 commented:
Please get [failure logs here](https://dev.azure.com/vcpkg/c1ee48cb-0df2-4ab3-8384-b1df5a79fe53/_apis/build/builds/105439/artifacts?artifactName=failure%20logs%20for%20x64-linux&api-version=7.1&%24format=zip) on `x64-linux`:
```
CMake Deprecation Warning at CMakeLists.txt:1 (CMAKE_MINIMUM_REQUIRED):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Warning (dev) at /mnt/vcpkg-ci/installed/x64-linux/share/pythoninterp/vcpkg-cmake-wrapper.cmake:10 (_find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

Call Stack (most recent call first):
  /mnt/vss/_work/1/s/scripts/buildsystems/vcpkg.cmake:811 (include)
  CMakeLists.txt:27 (FIND_PACKAGE)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at CMakeLists.txt:29 (MESSAGE):
  Python 2 is required.
```

valgur committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into update/suitesparse

MonicaLiu0311 commented:
Convert this PR to draft since there is no progress. Please ping us if this PR is ready for review again.

valgur commented:
The PR is broken by seemingly unrelated failures due to `python2` not being found in `graphqlparser`, which the recipe does not touch. I don't currently have any ideas regarding what broke it and how to fix it.

dg0yt commented:
This PR touches `scripts/ci.baseline.txt` so CI builds all ports which don't have cached artifacts. In the past, there were a number of baseline regressions, but most of them are fixed. `graphqlparser` was removed.
You can always push an empty commit to trigger CI. (`git commit -m CI --allow-empty`).

valgur committed:
CI

MonicaLiu0311 commented:
Please resolve the conflict.

valgur committed:
Merge remote-tracking branch 'origin/master' into update/suitesparse

# Conflicts:
#	ports/colmap/vcpkg.json
#	versions/baseline.json
#	versions/c-/colmap.json

valgur committed:
vdb

valgur committed:
[suitesparse-graphblas] only supported on x64 for windows

valgur committed:
vdb

valgur committed:
[SuiteSparse] update to v7.8.1

valgur committed:
vdb

valgur committed:
[suitesparse-spqr] add a missing 'supernodal' requirement for cholmod

valgur committed:
vdb

valgur committed:
[ceres] require matrixops feature from cholmod

valgur committed:
vdb

valgur committed:
[suitesparse-graphblas] move failures from "supports" to ci.baseline.txt

valgur committed:
vdb

MonicaLiu0311 commented on versions/c-/ceres.json:
```diff
@@ -1,7 +1,12 @@
+      "port-version": 6
+    },
+    {
+      "git-tree": "e628219999f68aaf9c3c33ba1a499e6a9b53e23d",
```
```suggestion
      "git-tree": "213619bf4451443bd69f784f9a7fa32fc2436e07",
```
Please restore the original data, that is the git-tree of 2.1.0#5, which should not be changed.

valgur committed:
fix ceres vdb

valgur committed:
fix g2o vdb

valgur committed:
[suitesparse-graphblas] adjust "supports"

valgur committed:
vdb

valgur committed:
[g2o] disable use of vendored CSparse

valgur committed:
vdb

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into update/suitesparse

dg0yt commented on ports/colmap/vcpkg.json:
```diff
@@ -25,6 +26,7 @@
     "gflags",
     "glew",
     "glog",
+    "metis",
```
This line is the only change to the colmap port. Is this intentional?

valgur replied:
It is. METIS is a direct dependency for COLMAP, which was previously provided transitively via SuiteSparse. The newer SuiteSparse versions vendor their METIS dependency with mangled symbols instead, so it needed to be explicitly listed now.

dg0yt commented on ports/suitesparse-amd/portfile.cmake:
```diff
@@ -0,0 +1,33 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO DrTimothyAldenDavis/SuiteSparse
+    REF v7.8.1
```
So `suiteparse-amd` uses the SuiteSparse repo, but a different versioning for the individual components?
- A comment here could make this clear. Actually it might be necessary to give a hint where to find the component version.
- Another comment in all ports might be necessary to remind maintainers that they should be updated together.

valgur replied:
The new versions of the sub-packages are helpfully listed under the release info on GitHub.

Nevertheless, since updating them by hand is quite tedious, I added a helper script to automate the update process. It already came in handy for the new v7.8.2 released just today.
I also included a comment mentioning this in all SuiteSparse manifest files.

dg0yt commented on ports/suitesparse-config/portfile.cmake:
```diff
@@ -0,0 +1,52 @@
+
+# Move SuiteSparseBLAS.cmake, SuiteSparsePolicy.cmake etc files
+file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/share")
+file(RENAME "${CURRENT_PACKAGES_DIR}/lib/cmake/SuiteSparse" "${CURRENT_PACKAGES_DIR}/share/SuiteSparse")
```
Would it work with lower-case `share/suitesparse`?
On windows and macos, standard filesystems are case insensitive, but on Linux it is a difference.

valgur replied:
Possibly? Would a lower-case version be preferable?

dg0yt replied:
IMO yes. Every port is expected to have a lower-case dir in `share`. If `suitesparse-config` creates  `share/SuiteSparse` before port `suiteparse` has the chance to create `share/suitesparse`, it starts to get inconsistent.
(OTOH one has to check the cmake files for hard-coded `SuiteSparse` pathes.)

valgur replied:
I renamed it to lower-case and it built without issues on Linux.

dg0yt commented on ports/theia/fix-external-dependencies.patch:
```diff
@@ -184,13 +243,17 @@ index a3ea187..7e9d912 100644
 -    "dependency: Ceres, please set CERES_INCLUDE_DIR.")
 -endif (CERES_FOUND)
-+find_package(Ceres REQUIRED)
++find_dependency(Ceres REQUIRED)
```
You can leave the original `find_package(... REQUIRED)`, or you can use the `find_dependency()` without `REQUIRED`.

valgur replied:
Done.

dg0yt commented on scripts/ci.baseline.txt:
```diff
@@ -1118,6 +1118,13 @@ stxxl:x86-windows=skip
+# _Interlocked* intrinsic functions are not available on x86
+suitesparse-graphblas:x86-windows=fail
+# "ninja: error: failed recompaction: Permission denied"
+suitesparse-graphblas:arm64-windows=fail
```
Parallel access?

valgur replied:
Perhaps, but I've never encountered this before and don't have a good idea how this might be fixed. Should I try disabling the parallel configuration of debug and release builds?

dg0yt replied:
> Should I try disabling the parallel configuration of debug and release builds?

Oh, you didn't disable it for this port but for other ports? That's the minimal step indeed. (The alternative is to look for step which touch files in the source dir.)

And if *this* happens in CI, it will also happen for users.

valgur replied:
Oh, nevermind, it already has parallel configure disabled...
I'll leave it as is for now then since it probably does not have a trivial fix.

valgur committed:
[SuiteSparse] add update_suitesparse.py script

valgur committed:
[SuiteSparse] update to v7.8.2

valgur committed:
[SuiteSparse] use lower-case `share/suitesparse` subdir

valgur committed:
vdb

valgur committed:
[theia] drop REQUIRED from `find_dependency(Ceres REQUIRED)`

valgur committed:
vdb

dg0yt commented:
The graphblas package made me study the package a little more. There are hard problems. (But maybe it is already broken now.)
- The configuration creates files in the source dir which capture per-build-type settings, incl. the compiler. One build type will see the wrong config unless running serializing config+build differently.
- Cross-builds with JIT (default, controlled by `compact`) rely on a cross-built external project for a host tool. In vcpkg, this should come from the host triplet instead.
- openmp off by default may not match upstreams recommendation printed in the config log:
~~~
CMake Warning at CMakeLists.txt:462 (message):
  WARNING: OpenMP was not found (or was disabled with GRAPHBLAS_USE_OPENMP).
  See the GraphBLAS user guide on the consequences of compiling GraphBLAS
  without OpenMP.  GraphBLAS will work but may not be thread-safe, since it
  relies on '#pragma omp flush' to ensure the work performed by one user
  thread is available to another, in GrB_wait.  If OpenMP is not in use, the
  thread-safety of GrB_wait becomes the responsibilty of the user application
  (perhaps through a pthreads construct).  Compiling GraphBLAS without OpenMP
  is not recommended for installation in a package manager (Linux,
  conda-forge, spack, brew, vcpkg, etc).
~~~

dg0yt commented on ports/suitesparse-graphblas/portfile.cmake:
```diff
@@ -0,0 +1,47 @@
+        -DSUITESPARSE_USE_STRICT=ON
+        -DSUITESPARSE_USE_FORTRAN=OFF
+        -DSUITESPARSE_DEMOS=OFF
+        -DGRAPHBLAS_JIT_ENABLE_RELOCATE=ON
```
Another note: This overrides upstream's default, and this is a point where upstream chooses a different default for APPLE.

dg0yt commented:
More on suitesparse-graphblas:

Postponing compilation to JIT, as modeled with feature "compact", saves a lot of build time. But this also implies that a significant part of the build is moved out of vcpkg, to a different time, maybe to a different machine (with the frozen absolute compiler paths and compilation flags). IMO this is fragile, at least from a more general vcpkg POV. I wonder if there are better models.

And I also wonder if "compact" is additive in a vcpkg feature sense, or if there should be "precompiled" instead, adding the extra build time and robustness when selected. With "core" being cheap, it would also be cheap to depend on the host port for the extra tool needed for JIT (if this dependency is feasible at all).

---

(Most of this might be relevant already for the current state of suitesparse, so it should not stop the PR. I just want to make sure that the split helps to move to an improved situation, maybe in following PRs.)

vicroms commented:
We discussed this PR in our last meeting, we think it is a good direction.

This is still pending thorough review, and we are a bit concerned about the upstream package names but will refer to whatever other package managers are doing.

vicroms commented on ports/suitesparse-cxsparse/vcpkg.json:
```diff
@@ -0,0 +1,31 @@
+  ],
+  "default-features": [
+    {
+      "name": "complex",
```
Default features are [difficult to turn off for users](https://learn.microsoft.com/vcpkg/concepts/default-features#disabling-default-features-in-manifest-mode), so we want to be careful with their inclusion. The fact that `complex` is not enabled/supported on all platforms makes me think this shouldn't be a default feature at all.

Does this feature comply with the "Default features should enable behaviors, not APIs" principle in our [contributor guidelines](https://learn.microsoft.com/vcpkg/contributing/maintainer-guide#default-features-should-enable-behaviors-not-apis)?

valgur replied:
I tried to match the behavior of the upstream project here, which disables complex support only for MSVC:
https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/v7.8.2/CXSparse/CMakeLists.txt#L48-L52

> Does this feature comply with the "Default features should enable behaviors, not APIs" principle in our contributor guidelines?

It only disables API-s: https://github.com/search?q=repo%3ADrTimothyAldenDavis%2FSuiteSparse+NCOMPLEX+NOT+path%3A%2F%5EGraphBLAS%5C%2FTest%5C%2F%2F+NOT+path%3A%2F%5ECXSparse%5C%2FMATLAB%5C%2F%2F&type=code

While the complex support is kind of important for the Cholesky decompositions this library deals with (since the returned matrices can be complex even if the input is real, if the matrix is not positive-definite), it does look like the feature could be disabled by default.

I would personally leave it enabled for consistency with the default behavior of the project.

vicroms replied:
If that's the case I would suggest always enabling `CXSPARSE_USE_COMPLEX` without a feature toggle.

vicroms commented on ports/suitesparse/update_suitesparse.py:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
```
This file should be moved to `scripts/suitesparse` rather than being part of the port.

vicroms commented on ports/suitesparse/vcpkg.json:
```diff
@@ -1,19 +1,131 @@
+    "suitesparse-spex"
+  ],
+  "default-features": [
+    "gpl"
```
Same here, users may not want to take GPL-licensed packages and having them by default will make it difficult to turn off, especially when `suitesparse` is depended on transitively.

valgur replied:
I would have left the `gpl` feature disabled by default, but it was effectively enabled before this update (i.e. all sub-packages, including GPL ones, were always enabled) and I tried to avoid breaking changes.

vicroms replied:
In the case of version updates, "breaking" changes are OK since users need to take action to update the library and vcpkg never updates without user request. This is going from `5.8.2` to `7.8.2`, that's two major versions which I believe already implies a level of breakage happening.

vicroms commented on ports/ceres/find-package-required.patch:
```diff
@@ -13,10 +13,11 @@ index 2d241fe..b79454d 100644
    # Check for SuiteSparse and dependencies.
 -  find_package(SuiteSparse 4.0 COMPONENTS CHOLMOD SPQR)
-+  find_package(suitesparse CONFIG REQUIRED)
++  find_package(CHOLMOD CONFIG REQUIRED)
++  find_package(SPQR CONFIG REQUIRED)
```
This is not a change request, I'm just curious as to how "transparent" these changes are:

* Do the changes in this PR require that we do this for all downstream consumers?
* Could we leave `find_package(suitesparse CONFIG REQUIRED)` and do the `find_package` updates in a follow-up PR?

MonicaLiu0311 commented:
@valgur

vicroms commented:
@valgur 

We are closing this PR due to inactivity. We like the changes here and would like to see them merged when you have time to address the requested changes (mainly moving the generation script to the `scripts` folder). Feel free to open a new PR when it's ready.
